src/analyzing/Constant_propagation.ml

```diff
@@ -749,45 +740,47 @@ let propagate_basic lang prog =
                             ( id,
                               {
                                 id_resolved = { contents = Some (kind, sid) };
+                                id_flags;
                                 _;
                               } )) ) );
                 _;
               },
               _,
               rexp ) ->
             let opt_svalue = eval env rexp in
-            (let is_private_class_field =
-               match Hashtbl.find_opt env.attributes (fst id, sid) with
-               | None -> false
-               | Some attrs ->
-                   List.exists is_private attrs && is_class_field env kind
-             in
-             let in_unique_constructor =
-               match ctx.in_class with
-               | None -> false
-               | Some cid ->
-                   ctx.in_constructor
-                   && has_just_one_constructor stats.class_stats (fst cid)
-             in
-             if
-               is_assigned_just_once stats.var_stats (H.str_of_ident id, sid)
-               (* restricted to prevent unexpected const-prop FPs *)
-               && ((is_lang env Lang.Python || is_lang env Lang.Ruby
-                  || is_lang env Lang.Php || is_js env)
-                   && is_global kind
-                  (* TODO: Add other Java-like OO languages, maybe Apex and C# ? *)
-                  || is_lang env Lang.Java && is_private_class_field
-                     && (ctx.in_static_block || in_unique_constructor))
-               && is_resolved_name kind sid
-             then
-               match opt_svalue with
-               | Some svalue -> add_constant_env id (sid, svalue) env
-               | None ->
-                   if
-                     Dataflow_svalue.is_symbolic_expr rexp
-                     && no_cycles_in_sym_prop sid rexp
-                   then add_constant_env id (sid, Sym rexp) env;
-                   ());
+            let is_private_class_field =
```

Comment:
The core change here is that we add `id_flags := IdFlags.set_final !id_flags;`, the rest is code reformatting.